### ISSUE TITLE
[Support Topic Request] Application Build Script Workflow

### ISSUE THREAD

**Issue by cdkharris** [UCL-ARC/hpc-support#4]
# Support Topic Request: Application Build Script Workflow

## Description:

This is a request to document the current workflow through which most research software is installed on the clusters and registered as a module. In the future it is planned to streamline this work with SPACK; however this workflow will still be needed to manage software that is not included in SPACK.

## Related to:

[rcps-modulefiles](https://github.com/UCL-RITS/rcps-modulefiles)
[rcps-buildscripts](https://github.com/UCL-RITS/rcps-buildscripts)

## Existing resources:

Documentation and examples in the above listed repositories.
https://github.com/UCL-ARC/arc-internal-documentation/blob/main/Advanced-Research-Computing/Technical-documentation/Research-Computing/General-service/Manually-building-software.md
This particular comment was a walkthrough of an updated version of an existing MPI install: https://github.com/UCL-ARC/rcps-buildscripts/issues/409#issuecomment-779966968 (after that comment it goes more complicated, but that comment has all the steps to look at).

## Recommendation:

Brian demonstrated the workflow to Fatima and myself yesterday. I took notes which I will use to write a first draft of an article in the GH wiki. Fatima and Brian will contribute for completeness and accuracy.

**Comment by cdkharris** [comment]
page started here: https://github.com/UCL-ARC/hpc-support/wiki/Research-Software-Build-Scripts
currently there is just a broad outline as a starting point.

**Comment by cdkharris** [comment]
TO-DO:

- [x] write an outline
- [x] flesh out the "development" section
- [x] flesh out the "testing" section
- [x] flesh out the "releasing" section
- [ ] review with team for clarity/accuracy/completeness

**Comment by ikirker** [comment]
I originally wrote the style in the current example with the idea that splitting everything up into stages would be useful for debugging, or even for processing with another tool somewhere down the line.

It never really was, and the function scopes and copy/pasting were more of an annoyance than anything else, so I switched to a simpler style incorporating some helper functions, including:

- `make_build_env`
- `make_module_v2`
- `manifest`
- `post_build_report`

These use the same variables as the current example, but some of the variables are never used explicitly in the resulting script, so they're a little more opaque.

They also assume you're running the script from the build scripts directory, so use a less arcane incantation for the includes.

You can see the result in e.g. <https://github.com/UCL-RITS/rcps-buildscripts/blob/master/emacs-26.3_install>.

I find these much simpler, but then I also wrote all the helper functions, so I know what's happening.

I'm happy to make the edits to the document (and the example, which I had completely forgotten about) to use this style, if people are happy for me to. I could also understand people wanting to keep aspects of the older style that keeps more things visible in the script.

There's also a small repo of simple Ansible-based tools for making things like updating the module repos on all the clusters easier: <https://github.com/UCL-RITS/rcps-ansible-tools>.

**Comment by cdkharris** [comment]
Yes please make these edits, I appreciate it. I had been looking at the VASP scripts when I was writing up the structure.

I think the style of the emacs script is easier to read, and more in keeping with what I expect to see in a bash script. 
`source includes/source_includes.sh` is much easier to understand, I agree.

**Comment by heatherkellyucl** [comment]
I updated a little on use of `ccspapp` - you can do the development as yourself or as it. Once done, you need to git pull the completed script as ccspapp.

**Comment by heatherkellyucl** [comment]
I tend to use [gromacs-2020.1_install](https://github.com/UCL-RITS/rcps-buildscripts/blob/master/gromacs-2020.1_install) as a fairly new style of buildscript, while [emacs-24.5_install](https://github.com/UCL-RITS/rcps-buildscripts/blob/master/emacs-24.5_install) is the oldest style (all the variables in caps at the top, a big if loop), and VASP with the pre_build, build, post_build, cleanup is Ian's first round of improvements and helper functions as he mentions.

It is worth knowing about the two older styles in case you need to do a new install based on one of those. It is probably worth gradually updating them as we go (providing they are not too complex).

**Comment by cdkharris** [comment]
Then it sounds like we should update the blank script to reflect the current preferred development style. I've also added references to these examples on the page.

**Comment by heatherkellyucl** [comment]
We have had a request for Gnuplot: https://github.com/UCL-RITS/rcps-buildscripts/issues/562 
This would be a good candidate for updating to new style and trying out the whole process.

I'd like @DanielChau128 and @ccaegul to see the process as well.

**Comment by cdkharris** [comment]
I've been developing the script for https://github.com/UCL-RITS/rcps-buildscripts/issues/562 . One thing I noticed is that the `module_maker` tool is (necessarily) a bit conservative when generating the modulefile. What I mean is that `module_maker` determined that `gcc-libs` and `compilers/gnu/4.9.2` should be prerequisites for `gnuplot`. But as far as I can tell it doesn't need them. The test plot came back fine without, and `ldd` indicates that while the executable is linked against some items in `gcc-libs`, the links use the absolute path, so I think it doesn't need the module to be loaded. I also checked that it works fine if a later version of `gcc-libs` is loaded.

This might not be the case if I had compiled it with support for qt5 or libcerf. So it's up to the developer to figure out what the application actually needs, but in my opinion it's better to not enforce prerequisites if they're not necessary.

**Comment by cdkharris** [comment]
It would be good to document the tests that we use to validate compiled software- in the case of `gnuplot` the test was:
```
module purge
module load gnuplot/6.0.rc2
gnuplot -e "set term png; test" > gnuplot_test.png
```

**Comment by heatherkellyucl** [comment]
Yup, module_maker will look at the modules loaded when building.

We usually have modules depend on `gcc-libs` unversioned and not on a compiler module, so they can be used when any compiler including Intel is loaded too.

Fortran programs link to specific versions of libgfortran so will need a specific gcc-libs.

**Comment by cdkharris** [comment]
a short description of how the permissions should be configured (1) for unrestricted applications and (2) for licensed software should be added
